### PR TITLE
Make image processors more general

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR undoes https://github.com/huggingface/transformers/pull/26965 and instead makes image processors more general, by removing the hardcoded "shortest_side" arguments, allowing to pass the following to image processors:

```
size = {"height": ..., "width": ...}
size = {"shortest_edge": ...}
```
Ideally it should also allow for the following (cc @amyeroberts)
```:
size = {"longest_edge": ...}
size = {"shortest_edge": ..., "longest_edge": ...}
```
Image processors should not be limited to only `{"shortest_edge": ...}` for instance, as `CLIPImageProcessor` is at the moment. The `size` argument is now always a dictionary containing [one of these 4 possibilities](https://github.com/huggingface/transformers/blob/7293fdc5b9cc809c2aa2ceb84f903ad47e5c06f0/src/transformers/image_processing_utils.py#L663); hence it would be great to support them all.

### CODE REVIEW

**Comment by amyeroberts** [comment]
Thanks for opening this PR! 

>Ideally it should also allow for the following (cc @amyeroberts)
> 
> size = {"longest_edge": ...}
> size = {"shortest_edge": ..., "longest_edge": ...}

As discussed on slack, enabling all image processors to accept `{"height": h, "width": w}` is a change I'm very much pro. Enabling all to accept `{"shortest_edge": x, "longest_edge": y}` isn't something I think we should as this behaviour isn't well-defined and varies between models.

**Comment by NielsRogge** [comment]
cc @ydshieh this is failing for KOSMOS-2 - it would be great to revert that change for KOSMOS-2 since it introduces a new attribute for legacy behaviour of size.

**Comment by ArthurZucker** [comment]
It did but it's also very convenient not to have to use copied from when the only difference was this, so reverting IMO is not the best solution. Let's rather adapt / make it compatible

**Comment by ydshieh** [comment]
Reverting #26965 means we well have to add `Kosmos2ImageProcessor` (while currently it is `CLIPImageProcessor` with `use_square_size=True`). Probably it will go without problem, but still kind of breaking changes.

**Comment by NielsRogge** [comment]
@ydshieh would it also be possible to just set size to `{"height": ..., "width" ...}`? With this PR, image processors are made more general, making sure square sizes are supported

**Comment by ydshieh** [comment]
Hi, could you give more details on

>  set size to {"height": ..., "width" ...}

(where should I set this, for example.)

**Comment by ydshieh** [comment]
We can keep your backwards compatibility code, and I could try to open Hub PRs to update them.
Let's also hear what the core maintainers say of course.

**Review by ArthurZucker — COMMENTED** [review]
Looks like a better and more general solution so in favor, I'll let @amyeroberts review when she comes back unless it's urgent!  😉

**Comment by NielsRogge** [comment]
It's required to merge #27718

File: src/transformers/models/test.py
Comment (amyeroberts) [inline src/transformers/models/test.py:1]: This should be removed

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding! 

The test file will need to be removed before merging - otherwise looks good!

**Review by amyeroberts — APPROVED** [review]

**Comment by ydshieh** [comment]
Merge now as full (slow) CI on 4 models touched by this PR pass

**Comment by ArthurZucker** [comment]
cc @younesbelkada we have a regression on this

**Pull request merged**
